### PR TITLE
Support broadcasting just lora weights and drop support for merged weights

### PR DESCRIPTION
**Pull request by Jackmin801**
> [!NOTE]
> Adds adapter-only filesystem weight broadcasting and runtime LoRA loading, wires LoRA names through orchestrator/scheduler, removes merged-weight path, and updates tests/CI accordingly.
> 
> - **LoRA broadcasting and loading**:
>   - Add `adapter_only` option to weight broadcast configs (`trainer/rl/config.py`); filesystem broadcaster now saves only adapters and writes `adapter_config.json`; NCCL path disallows adapter-only.
>   - Introduce `get_adapter_state_dict` and adapter filename handling in `trainer/weights.py`; drop merged-LoRA gathering path; simplify `gather_weights_on_master`.
>   - Pass `lora_config` into broadcaster setup and persist with adapter saves.
> - **Orchestrator & Scheduler**:
>   - Add `OrchestratorConfig.lora_name` and thread through `Scheduler`; use LoRA name as `model_name` after updates.
>   - `update_weights` now accepts `lora_name` and uses new `/v1/load_lora_adapter`; avoid base reload when using LoRA.
> - **Inference (vLLM server)**:
>   - Inject hack to bypass adapter already-loaded checks to allow in-place LoRA updates.
> - **Trainer**:
>   - Broadcast weights each step using `adapter_only` flag; checkpoint manager unchanged but aligned with new adapter flow.
> - **Tests/CI**:
>   - Add integration tests for LoRA RL with dynamic adapter loading; separate LoRA tests in CI; capture vLLM/stdout logs for debugging.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 7857428f9e02b9314087b5f609c41f135554cc2c. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: .github/workflows/gpu_tests.yaml [hunk 1]
```diff
@@ -76,4 +76,6 @@ jobs:
           GITHUB_REF_NAME: ${{ github.ref_name }}
           GITHUB_HEAD_REF: ${{ github.head_ref }}
           PYTEST_OUTPUT_DIR: /tmp/outputs
-        run: uv run pytest tests/integration -m gpu
+        run: |
+          uv run pytest tests/integration -m gpu --ignore=tests/integration/lora/test_rl.py
```
Comment (mikasenghaas) [inline .github/workflows/gpu_tests.yaml:80]: why this separation?

File: src/prime_rl/orchestrator/config.py [hunk 2]
```diff
@@ -520,6 +520,13 @@ class OrchestratorConfig(BaseSettings):
 
     seed: Annotated[int | None, Field(description="Random seed for the orchestrator.")] = 42
 
+    lora_name: Annotated[
```
Comment (mikasenghaas) [inline src/prime_rl/orchestrator/config.py:523]: this is what we'd use for identifying the lora tenants right?

File: src/prime_rl/orchestrator/scheduler.py [hunk 3]
```diff
@@ -162,10 +165,14 @@ async def update_policy(self):
...
             update_weights_start_time = time.perf_counter()
             await update_weights(
-                self.admin_clients, get_step_path(get_broadcast_dir(self.config.output_dir), next_ckpt_step)
+                self.admin_clients,
+                get_step_path(get_broadcast_dir(self.config.output_dir), next_ckpt_step),
+                lora_name=self.lora_name,
             )
             self.update_weights_time = time.perf_counter() - update_weights_start_time
             self.logger.debug(f"Updated weights to step {next_ckpt_step} in {self.update_weights_time:.2f}s")
+            if self.lora_name is not None:
```
Comment (mikasenghaas) [inline src/prime_rl/orchestrator/scheduler.py:174]: why this here?

File: .github/workflows/gpu_tests.yaml (hunk 1 above)
Comment (mikasenghaas) [reply .github/workflows/gpu_tests.yaml:80]: ah nvm, makes sense. do we have an idea what the additional overhead is for ci time?

**Review by mikasenghaas — COMMENTED** [review]
very clean, i love it. nice that we decouple the full vs adapter only weight saving as well, has been bugging me for a while how entangled these were

File: .github/workflows/gpu_tests.yaml (hunk 1 above)
Comment (Jackmin801) [reply .github/workflows/gpu_tests.yaml:80]: yea they use vllm server with different args. The non lora one supports update weights but not load lora, the lora one supports load lora but not update weights

Comment (Jackmin801) [reply .github/workflows/gpu_tests.yaml:80]: its additional ci time ~300s

File: src/prime_rl/orchestrator/config.py (hunk 2 above)
Comment (Jackmin801) [reply src/prime_rl/orchestrator/config.py:523]: yep!

File: src/prime_rl/orchestrator/scheduler.py (hunk 3 above)
Comment (Jackmin801) [reply src/prime_rl/orchestrator/scheduler.py:174]: reset weights doesnt work with the lora vllm server

File: .github/workflows/gpu_tests.yaml (hunk 1 above)
Comment (samsja) [reply .github/workflows/gpu_tests.yaml:80]: > its additional ci time ~300s
might be worth it no ? Very imporant the we don't break lora as we move on imo.

But maybe we can wait to enable it for when we have a better CI setup with mode worker

File: src/prime_rl/trainer/rl/config.py [hunk 4]
```diff
@@ -52,6 +52,7 @@ class FileSystemWeightBroadcastConfig(BaseModel):
     """Configures the weight broadcast."""
 
     type: Literal["filesystem"] = "filesystem"
+    adapter_only: Annotated[bool, Field(description="Whether to save LoRA adapters only for weight broadcast.")] = False
```
Comment (samsja) [inline src/prime_rl/trainer/rl/config.py]: woulnd't be clearer to call it lora_only ?

File: src/prime_rl/trainer/rl/config.py [hunk 5]
```diff
@@ -62,6 +63,7 @@ class NCCLWeightBroadcastConfig(BaseModel):
     """Configures the NCCL broadcast."""
 
     type: Literal["nccl"] = "nccl"
+    adapter_only: Annotated[bool, Field(description="Whether to save LoRA adapters only for weight broadcast.")] = False
```
Comment (samsja) [inline src/prime_rl/trainer/rl/config.py]: @mikasenghaas probably will want to do a base config for the shared field for broadcast, I don't mind keeping it like this for now tho

File: src/prime_rl/trainer/rl/config.py (hunk 4 above)
Comment (Jackmin801) [reply src/prime_rl/trainer/rl/config.py]: adapter is more general though. we might want to support other peft methods in the future
